sky/skylet/providers/kubernetes/node_provider.py

```diff
@@ -205,14 +204,58 @@ def create_node(self, node_config, tags, count):
         # Wait for all pods to be ready, and if it exceeds the timeout, raise an
         # exception. If pod's container is ContainerCreating, then we can assume
         # that resources have been allocated and we can exit.
-
         start = time.time()
         while True:
             if time.time() - start > self.timeout:
-                raise config.KubernetesError(
-                    'Timed out while waiting for nodes to start. '
-                    'Cluster may be out of resources or '
-                    'may be too slow to autoscale.')
+                for new_node in new_nodes:
+                    pod_status = new_node.status.phase
+                    pod_name = new_node._metadata._name
+                    events = kubernetes.core_api().list_namespaced_event(
+                        self.namespace,
+                        field_selector=(f'involvedObject.name={pod_name},'
+                                        'involvedObject.kind=Pod'))
+                    # Events created in the past hours are kept by
+                    # Kubernetes python client and we want to surface
+                    # the latest event message
+                    events_desc_by_time = \
+                        sorted(events.items,
+                        key=lambda e: e.metadata.creation_timestamp,
+                        reverse=True)
+                    for event in events_desc_by_time:
+                        if event.reason == 'FailedScheduling':
+                            event_message = event.message
+                            break
+                    timeout_err_msg = (
+                        'Timed out while waiting for nodes to start. '
+                        'Cluster may be out of resources or '
+                        'may be too slow to autoscale.')
+                    lack_resource_msg = (
+                        'Insufficient {resource}. Other SkyPilot tasks or pods on '
+                        'the cluster may be using resources. Check resource usage '
+                        'by running `kubectl describe nodes`.')
+                    if event_message is not None:
+                        if pod_status == 'Pending':
+                            if 'Insufficient cpu' in event_message:
+                                raise config.KubernetesError(
+                                    lack_resource_msg.format(resource='CPUs'))
+                            if 'Insufficient memory' in event_message:
+                                raise config.KubernetesError(
+                                    lack_resource_msg.format(resource='memory'))
+                            gpu_lf_keys = [lf.get_label_key() \
+                                for lf in kubernetes_utils.LABEL_FORMATTER_REGISTRY]
+                            # Confirms if the nodeSelector in the pod spec is
+                            # used for GPU scheduling.
+                            for label_key in new_node.spec.node_selector.keys():
```

Comment:
This fails with `AttributeError: 'NoneType' object has no attribute 'keys'`. See [logs and reproduction here](https://gist.github.com/romilbhardwaj/defe0b02ceb06d4eaeb203b4e57db536) (I used kubectl taint to prevent scheduling). Note that `new_node.spec.node_selector` may be None on CPU only tasks.

Reply:
More generally, it might be a good idea to wrap all this special message handling in a try catch block, and if any step fails then raise a generic error (+ the error from the try catch block, if SKYPILOT_DEBUG=1).